### DISCUSSION TITLE
A Dagger + Tilt Experiment

### DISCUSSION THREAD

**Discussion by gtabrar** [dagger/dagger#8044] [General]
I wanted to share some feedback from one of our engineers, who put together a basic test for running Tilt with Dagger. For the test we used Docker Engine v26.0.0 in a GCP virtual machine running Ubuntu 22.04, using Dagger v0.12.0 and the [k3s v0.0.4](https://daggerverse.dev/mod/github.com/marcosnils/daggerverse/k3s) Daggerverse module.

### Dagger was successfully able to:

- Start a k3s cluster as a Service
- Create an Ubuntu 22.04 build container, configured to use the k3s cluster
- Run Tilt inside that build container

### Tilt successfully stood up the defined service using the configured k3s cluster. Tilt was around 90% functional in this test:

- The build container was not privileged and Tilt could not connect to Docker
- Kubernetes log streaming did not work: k3s returns a 502 error from the Kubernetes API gateway
- There is some question if the tilt ci command will work as we need

### This tests did not cover:

- Testing using Kind. The Kind Daggerverse module did not build successfully.
- Running tests using Tilt (E.g. as frs-tilt runs Venom)
- Running the pipeline inside the existing FBS GKE environment (yet)
- How the pipeline will work in practice E.g. How do we make build-time resources available to Tilt during testing?

#### Async Service Binding
I had some difficulty getting the k3s configuration to work: Dagger would successfully start k3s but the build container would contain a Kube configuration that did not work and referenced a k3s cluster that did not exist.

It took a little time to realize that the build container had a configuration file that contained the details for the previous k3s cluster (E.g. from the last time I ran the pipeline).

It appears that Dagger has no way to create a hard dependency between a container and the service that is bound to it: Dagger will not wait for the Service to start & be healthy before creating the Container that has a ServiceBinding on it, hence the container would start and retrieve the previous Kube configuration from the Dagger cache, all while Dagger was starting k3s as a service (and over-writing the Kube configuration with the new details after the container was started).

I worked around this partially by forcing Dagger to create a "dummy" Container with the ServiceBinding against it before creating the build Container proper, but this is quite a hack (and may still be a race condition).

There may be some way to make this work correctly with E.g. a readiness check on the Service that would cause Dagger to wait until it was up & healthy but I don't see any obvious way to do this.

#### Kind
Tilt recommended testing with [Kind](https://kind.sigs.k8s.io/). In theory a [Daggerverse module](https://daggerverse.dev/mod/github.com/aweris/daggerverse/kind) exists that would create a Kind cluster. In practice this module has not been updated in nearly a year and did not compile:

```
✘ ModuleSource.asModule: Module! 9.9s
! failed to create module: select: failed to update module dependencies: failed to initialize dependency modules: failed to initialize dependency module: select: failed to initialize module: failed to call module "kind" to get functions: call constructor: process "go build -o /runtime ." did not complete successfully: exit code: 1
Error: failed to generate code: input: moduleSource.withContextDirectory.withDependencies.asModule resolve: failed to create module: select: failed to update module dependencies: failed to initialize dependency modules: failed to initialize dependency module: select: failed to initialize module: failed to call module "kind" to get functions: call constructor: process "go build -o /runtime ." did not complete successfully: exit code: 1
```

Stderr:
```
# kind
./kind.go:23:13: undefined: Optional
./kind.go:43:7: undefined: Optional
./kind.go:45:13: undefined: Optional
./kind.go:58:45: undefined: Opt
./kind.go:73:7: undefined: Optional
./kind.go:75:13: undefined: Optional
./kind.go:176:11: undefined: Optional
./kind.go:267:69: dag.Host undefined (type *dagger.Client has no field or method Host)
```

#### Tilt notes
Tilt can interact with both Docker & Kubernetes directly. Without a privileged container, Tilt is (naturally) not able to connect to Docker. Any Tilt features that rely on Docker acces will not work.

It's unclear from the Dagger documentation if there is any way to create a privileged container with access to Docker, nor is it clear if we would want to do that.

#### Log Streaming
Any attempt to stream logs from a kubelet fails with a 502 error from the Kubernetes API gateway E.g.

```
Error streaming frs-mysql-0 logs: Get "https://10.87.0.96:10250/containerLogs/default/frs-mysql-0/mysql?follow=true&sinceTime=2024-07 23T13%3A20%3A41Z": proxy error from 10.87.0.96:6443 while dialing 10.87.0.96:10250, code 502: 502 Bad Gateway
```

This problem affects both Tilt (during service creation) and kubectl (from a terminal inside the build container). It's unclear if this is a general problem with k3s or something more specific to this configuration.

#### Tilt CI
It's unclear if the tilt ci command actually works the way that we would need it to for this use case. The [Tilt blog](https://blog.tilt.dev/2020/04/16/how-to-not-break-server-startup.html) describes the command, but this doesn't really fit the way that E.g. frs-tilt runs tests (A Kubernetes job that creates a container that runs the Venom test suite). A current [open issue](https://github.com/tilt-dev/tilt/issues/6315) suggests this wouldn't work.

Tilt suggests using local_resource to run tests, which would work for users who use test frameworks that run outside of the cluster. In theory this should work fine with this model, but we would need to test this. In addition we might want to do more testing to see if there is a reliable way to trigger frs-tilt style tests inside the cluster (I.e. automatically started by tilt ci).

##### Tilt CI Behaviour
In theory the tilt ci command should be non-interactive but in practice there were situations where Tilt would stop and block the running Dagger pipeline. This often happened when Tilt encountered an early error E.g. if it could not connect to the Kubernetes cluster, where it would stop with a message to "Restart Tilt once you have fixed the problem" and then stop without exiting.

### Other notes
These notes relate to my experience using Dagger.

1. The Daggerverse modules are generally of poor quality. Many appear to have been developed briefly and then abandoned. I'd have concerns about using any of these modules in production right now.

2. There are sometimes multiple modules with the same name E.g. [k3s](https://daggerverse.dev/mod/github.com/marcosnils/daggerverse/k3s) v's [k3s](https://daggerverse.dev/mod/github.com/aweris/daggerverse/k3s) (There appear to be five called "docker")
The documentation for many of these modules is poor E.g. the k3s module I use doesn't even get the module namespace correct (K3s in the docs. v's K3S in reality)

3. Managing Daggerverse modules is opaque. You can install a module (dagger install) but that's it. Once a module is installed you're on your own, with no way to list the installed modules or uninstall one.
You can install two modules that shadow each other (E.g. they use the same namespace): I accidentally did this (See above re. the two k3s modules) and Dagger didn't complain beside random compile failure when the "other" module won the race.

4. The documentation is sometimes non-existent or poor
For example I can find nothing that documents what methods (E.g. From(), With…() etc.) are available for the core types (E.g. Container, Service etc.), or even what the canonical set of core types are.

5. The async nature of the pipeline can be confusing
The asynchronous nature of the DAG combined with opaque caching combined with the inability to control that DAG (E.g. wait for X to happen before Y) can be confusing (See Async Service Binding above).

**Comment by jpadams** [comment]
First of all, thank you for spending time on doing these tests and writing it all up. This feedback is super valuable for us.

We'd love to work with you to make this POC work as you'd expect (get to 100%). In particular, we'd like to dig into networking and health check challenges you had with Service containers (more polish planned there). Let us know if you'd like to pair up on it.

Your feedback is well-taken and super appreciated. Some more notes below.

We're rapidly iterating on Dagger and some features are still converging. In particular `v0.12.0` introduced some features and breaking changes to the previous API that are still being perfected (https://github.com/sipsma/dagger/blob/main/.changes/v0.12.0.md). We're working on a `v0.12.4`  release today!

While some modules in the Daggerverse are from folks on the Dagger team, most are a community effort and sometimes a person's side project, so many are experimental or not fully up to date. Most folks are happy to learn that someone is using their module and will take nudges and PRs. I see movement on both the `kind` and `k3s` modules you mentioned, for example.

In the near term, we want to float the most up-to-date and best-documented modules to the top. Most modules should “just work” going forward (https://github.com/dagger/dagger/pull/7751. The old `kind` module suffered from a syntax retired longer ago). On the topic of module documentation, there is already a feature for more more detailed examples that we need to polish and promote more (https://daggerverse.dev/mod/github.com/marcosnils/daggerverse/gptools@053004f8c41acc294c4d861bcc15c8f60c931057). Longer term, we have a roadmap of features around module discovery, documentation, code quality, and safety (including supported modules).

Your feedback on Service containers is super helpful since Service containers were introduced before Modules and there are still some nuances to get right. We have some issues open, but want to incorporate your specific experience. Comments welcome!
https://github.com/dagger/dagger/issues/8037
https://github.com/dagger/dagger/issues/7426

We're going to continue to improve the Dagger docs including more guides on API/architecture and lazy/async execution, as well as a docs restructure underway that we hope will make it easier to get started. We'll be sure to make the API reference docs more visible, since they are a bit buried today:
https://docs.dagger.io/manuals/developer
Go, Python, TypeScript SDK references
(see also https://docs.dagger.io/manuals/developer/ide-integration/)

On a related note, some of us have talked about having the Core API documented in the Daggerverse as well. This goes along with the effort to make the Core API accessible from the CLI as well: https://github.com/dagger/dagger/issues/6947#issuecomment-2245024282
We'd love to hear your thoughts on that.

Again, thanks for all the feedback! It helps to make Dagger better for everyone. 🚀
